artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerTarget.java

```diff
@@ -38,51 +35,136 @@
 
-   public static final SimpleString INTERNAL_ID_EXTRA_PROPERTY = SimpleString.toSimpleString(INTERNAL_ID.toString());
-
    private static final Logger logger = Logger.getLogger(AMQPMirrorControllerTarget.class);
 
-   final ActiveMQServer server;
+   private static ThreadLocal<MirrorController> controllerThreadLocal = new ThreadLocal<>();
+
+   public static void setControllerTarget(MirrorController controller) {
+      controllerThreadLocal.set(controller);
+   }
+
+   public static MirrorController getControllerTarget() {
+      return controllerThreadLocal.get();
+   }
+
+   class ACKMessageOperation extends TransactionOperationAbstract implements IOCallback, Runnable {
+
+      Delivery delivery;
+
+      void reset() {
+         this.delivery = null;
+      }
+
+      ACKMessageOperation setDelivery(Delivery delivery) {
+         this.delivery = delivery;
+         return this;
+      }
+
+      @Override
+      public void run() {
+         if (logger.isTraceEnabled()) {
+            logger.trace("Delivery settling for " + delivery + ", context=" + delivery.getContext());
+         }
+         delivery.disposition(Accepted.getInstance());
+         settle(delivery);
+         connection.flush();
+         AMQPMirrorControllerTarget.this.ackMessageMpscPool.release(ACKMessageOperation.this);
+      }
+
+      @Override
+      public void beforeCommit(Transaction tx) throws Exception {
+      }
+
+      @Override
+      public void afterCommit(Transaction tx) {
+         done();
+      }
+
+      @Override
+      public void done() {
+         connection.runNow(ACKMessageOperation.this);
+      }
+
+      @Override
+      public void onError(int errorCode, String errorMessage) {
+         logger.warn(errorMessage + "-"  + errorMessage);
+      }
+   }
+
+   // in a regular case we should not have more than amqpCredits on the pool, that's the max we would need
+   private final MpscPool<ACKMessageOperation> ackMessageMpscPool = new MpscPool<>(amqpCredits, ACKMessageOperation::reset, () -> new ACKMessageOperation());
 
    final RoutingContextImpl routingContext = new RoutingContextImpl(null);
 
-   Map<SimpleString, Map<SimpleString, QueueConfiguration>> scanAddresses;
+   final BasicMirrorController<Receiver> basicController;
+
+   final ActiveMQServer server;
+
+   final DuplicateIDCache duplicateIDCache;
+
+   private final IDSupplier<MessageReference> referenceIDSupplier;
 
    public AMQPMirrorControllerTarget(AMQPSessionCallback sessionSPI,
                                      AMQPConnectionContext connection,
                                      AMQPSessionContext protonSession,
                                      Receiver receiver,
                                      ActiveMQServer server) {
       super(sessionSPI, connection, protonSession, receiver);
+      this.basicController = new BasicMirrorController(server);
+      this.basicController.setLink(receiver);
```

Comment:
I would make this part of the construction, and verify the remote broker id at this point, failing if it isnt found given there are uses later that rely on it (but dont really check it).

I would probably add an offered+desired link capabilities handshake check too to verify that it is a peer on the other side that understands the new mirroring setup.

Reply:
@gemmellr yes.. I have seen this comment already... I'm still thinking on how I would / should address this.... leave it with me for a day or so before I can comment back on this one.

Reply:
I am now verifying the broker ID as part of the callback on the remote open.

The source will call sender.open(), set a callback on the link through link.setContext();

The callback will then be called on the remote-open, and if the ID is not specified, it will issue an error, and do a regular retry through the connection.